docs/proposals/caching.adoc

```diff
@@ -0,0 +1,164 @@
+
+.TODO figure out where/how to implement workspace cache disablement. 
+
+To ensure that mainline builds cannot be impacted by workspace caching ods-pipeline ensures the following:
+
+- Build pipelines for simple branch names will disable workspace caching.
+- When opening a pull request a non workspace cached build is triggered if the current build of this commit was workspace cashed (or always).
+- Pushing commits when a pull request is already open will disable workspace caching. 
+
+ods-pipeline should support disable build with a special tag in a commit comment.
+ 
+Furthermore only build tasks for which workspace caching provides a significant performance gain will support workspace caching. Candidates are node and python related build tasks.
+
+* The workspace directory is where the source code is checked out. This is the current working directory when a build task starts running. Without caching the workspace directory will contain the checked out sources. A task will build at the workspace directory or below if `WORKING_DIR` is not set to `"."`. 
+
+With sufficient caching space available the workspace directory persists from a prior build if in `ods.yaml` workspace caching is enabled. By default workspace caching is not enabled.
+
+Without caching enabled `ods-start` checks out sources via `git-init` unchanged from the current implementation. With caching and if the workspace already exists, then the sources are updated via `git pull` instead and other files will be left alone, so that the build can take advantage of already installed dependencies for example.  
+
+A new pipeline task parameter is introduced:
+
+* `cache-workspace-require-space-mb` a number defaulting to `0`. This is the number of MB required for the cache workspace for this build task. If greater 0 caching will be enabled space permitting and if not disabled as described above. 
+
+Task implementation are expected to not pass the parameter to the build scripts as these should not be affected by this.
+
+When `ods-start` determines that workspace caching is available, it adds the following file:
+
+* `$WORKING_DIR/.ods/git-dir-commit-sha` which contains the git commit sha of the working directory (no whitespace or newlines). Build tasks can use this to avoid rebuilding when there were no changes in their working dir.
+
+Build tasks supporting caching will also be adjusted to
+
+- Log how long build commands which may be long running take in seconds. 
+
+- Log available disk space well. This should be implemented so that one can see what sized to require.
+
+=== Global cache
+
+Only build tasks for which a global cache provides a significant performance gain will support global caching. At the moment the primary candidates is the go language. Nexus does not support go and we have no similar go dependency artifacts manager in place.
```

Comment:
gradle will also benefit. Normally because it always download the wrapper, unless the url that points to the gradlew wrapper is being updated to point to nexus down (https://github.com/opendevstack/ods-core/issues/919). However to avoid this problem, I added to the gradle task an instruction to download the gradlew wrapper so that it lands in the gradle cache folder (https://github.com/opendevstack/ods-pipeline/blob/0ae2f9bf95d2864f8fb2ff5012f6ff9085f89b33/build/package/Dockerfile.gradle-toolset#L22). This has the limitation that just only one version of the gradle wrapper is available.